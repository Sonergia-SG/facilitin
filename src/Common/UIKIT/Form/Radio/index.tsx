import React, { useState } from 'react';

interface Props {
  id: string;
  name: string;
  checked: boolean;
  disabled?: boolean;
  value: string;
  customColor?: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

type ComputeStrokeColor = (
  disabled: boolean | undefined,
  checked: boolean,
  hover: boolean,
  color: string
) => string;

const computeStrokeColor: ComputeStrokeColor = (disabled, checked, hover, color) => {
  if (disabled) return checked ? color : 'grey';

  return checked || hover ? color : 'grey';
};

type ComputeOpacity = (
  disabled: boolean | undefined,
  hover: boolean,
  trigerChange: boolean
) => number;

const computeOpacity: ComputeOpacity = (disabled, hover, trigerChange) => {
  if (disabled) return 0.5;

  return hover && trigerChange ? 0.7 : 1;
};

const Radio = ({
  checked, disabled, onChange, value, id, name, customColor,
}: Props) => {
  const [hover, setHover] = useState(false);

  const color = customColor || '#1fb5ad';
  const fillColor = checked ? color : 'grey';
  const strokeColor = computeStrokeColor(disabled, checked, hover, color);

  const trigerChange = !checked && !disabled;

  return (
    <div
      role="button"
      tabIndex={0}
      style={{ outline: 'none', position: 'relative', cursor: disabled ? 'auto' : 'pointer' }}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
    >
      <svg
        version="1.1"
        baseProfile="full"
        xmlns="http://www.w3.org/2000/svg"
        width={20}
        height={20}
        style={{
          position: 'absolute',
          opacity: computeOpacity(disabled, hover, trigerChange),
        }}
      >
        <circle cx={10} cy={10} r={8} strokeWidth={2} stroke={strokeColor} fill="#0000" />
        {checked && <circle cx={10} cy={10} r={4} fill={fillColor} />}
      </svg>
      <input
        style={{
          width: 20,
          height: 20,
          position: 'absolute',
          opacity: 0,
        }}
        type="radio"
        id={id}
        name={name}
        checked={checked}
        value={value}
        disabled={disabled}
        onChange={onChange}
      />
    </div>
  );
};

export default Radio;
