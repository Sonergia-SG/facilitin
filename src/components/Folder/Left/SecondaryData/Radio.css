.SecondaryData-Radio-Label {
  margin-bottom: 5px;
  margin-top: 0;
  font-size: 13px;
  font-weight: 700;
  color: #767676;
}

.SecondaryData-Radio {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 5px 2px;
  border-radius: 2px;
  transition: background-color 0.2s ease;
}

.SecondaryData-Radio:hover {
  background-color: #f8f8f8;
}

.SecondaryData-Radio > label {
  flex: 1;
  text-align: right;
  cursor: pointer;
  color: #767676;
  font-size: 13px;
  font-weight: 400;
}
