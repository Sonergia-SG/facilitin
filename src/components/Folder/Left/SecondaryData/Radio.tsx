import React, { ChangeEvent } from 'react';
import idx from 'idx';

import { FormFieldRadio } from './types';
import UIRadio from '../../../../Common/UIKIT/Form/Radio';
import { FolderFull } from '../../../../store/reducer/entities/types';
import { FolderPendingItem } from '../../../../store/reducer/views/folder/types';
import {
  folderUpdateMoaValue,
  folderUpdateMoeValue,
  folderUpdateSiteValue,
} from '../../../../store/actions';

import './Radio.css';

type GenericUpdate = (idDpOperation: number, key: string, value: string) => void;

interface Props {
  data: FormFieldRadio;
  idDpOperation: number;
  disabled: boolean;
  dossierprime: FolderFull;
  pending?: FolderPendingItem;
  pendingKey: 'moe' | 'moa' | 'site';
  update:
  | GenericUpdate
  | typeof folderUpdateMoaValue
  | typeof folderUpdateMoeValue
  | typeof folderUpdateSiteValue;
}

const Radio = ({
  data,
  update,
  idDpOperation,
  pendingKey,
  pending,
  dossierprime,
  disabled,
}: Props) => {
  const v = idx(pending, _ => _[pendingKey][data.key]);
  const originalValue = data.value || dossierprime[data.key];
  const cleanOriginalValue = originalValue === null || originalValue === undefined ? '' : originalValue;

  // const isEdited = v !== undefined;

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    update(idDpOperation, data.key, e.target.value);
  };

  const selected = typeof v === 'string' ? v : cleanOriginalValue;

  return (
    <div>
      <p className="SecondaryData-Radio-Label">{data.label}</p>
      {data.values.map(value => (
        <div className="SecondaryData-Radio" key={value.key}>
          <UIRadio
            checked={selected === value.key}
            id={value.key}
            name={value.key}
            onChange={handleChange}
            disabled={disabled}
            value={value.key}
          />
          <label htmlFor={value.key}>{value.value}</label>
        </div>
      ))}
    </div>
  );
};

export default Radio;
