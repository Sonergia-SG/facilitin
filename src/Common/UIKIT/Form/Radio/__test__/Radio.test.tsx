import React from 'react';
import { shallow } from 'enzyme';

import Radio from '../index';

describe('Radio', () => {
  it('render checked correctly', () => {
    const wrapper = shallow(
      <Radio checked id="test" name="test" value="test" onChange={() => {}} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('render not checked correctly', () => {
    const wrapper = shallow(
      <Radio checked={false} id="test" name="test" value="test" onChange={() => {}} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('render checked and disabled correctly', () => {
    const wrapper = shallow(
      <Radio checked disabled id="test" name="test" value="test" onChange={() => {}} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('render not checked and disabled correctly', () => {
    const wrapper = shallow(
      <Radio checked={false} disabled id="test" name="test" value="test" onChange={() => {}} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('render checked and hover correctly', () => {
    const wrapper = shallow(
      <Radio checked id="test" name="test" value="test" onChange={() => {}} />,
    );

    wrapper.simulate('mouseEnter');

    expect(wrapper).toMatchSnapshot();
  });

  it('render not checked and hover correctly', () => {
    const wrapper = shallow(
      <Radio checked={false} id="test" name="test" value="test" onChange={() => {}} />,
    );

    wrapper.simulate('mouseEnter');

    expect(wrapper).toMatchSnapshot();
  });

  it('render checked, disabled and hover correctly', () => {
    const wrapper = shallow(
      <Radio checked disabled id="test" name="test" value="test" onChange={() => {}} />,
    );

    wrapper.simulate('mouseEnter');

    expect(wrapper).toMatchSnapshot();
  });

  it('render not checked, disabled and hover correctly', () => {
    const wrapper = shallow(
      <Radio checked={false} disabled id="test" name="test" value="test" onChange={() => {}} />,
    );

    wrapper.simulate('mouseEnter');

    expect(wrapper).toMatchSnapshot();
  });
});
