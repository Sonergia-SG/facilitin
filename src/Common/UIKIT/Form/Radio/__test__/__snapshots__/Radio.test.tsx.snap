// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio render checked and disabled correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "auto",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 0.5,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="#1fb5ad"
      strokeWidth={2}
    />
    <circle
      cx={10}
      cy={10}
      fill="#1fb5ad"
      r={4}
    />
  </svg>
  <input
    checked={true}
    disabled={true}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;

exports[`Radio render checked and hover correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "pointer",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 1,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="#1fb5ad"
      strokeWidth={2}
    />
    <circle
      cx={10}
      cy={10}
      fill="#1fb5ad"
      r={4}
    />
  </svg>
  <input
    checked={true}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;

exports[`Radio render checked correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "pointer",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 1,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="#1fb5ad"
      strokeWidth={2}
    />
    <circle
      cx={10}
      cy={10}
      fill="#1fb5ad"
      r={4}
    />
  </svg>
  <input
    checked={true}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;

exports[`Radio render checked, disabled and hover correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "auto",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 0.5,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="#1fb5ad"
      strokeWidth={2}
    />
    <circle
      cx={10}
      cy={10}
      fill="#1fb5ad"
      r={4}
    />
  </svg>
  <input
    checked={true}
    disabled={true}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;

exports[`Radio render not checked and disabled correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "auto",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 0.5,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="grey"
      strokeWidth={2}
    />
  </svg>
  <input
    checked={false}
    disabled={true}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;

exports[`Radio render not checked and hover correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "pointer",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 0.7,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="#1fb5ad"
      strokeWidth={2}
    />
  </svg>
  <input
    checked={false}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;

exports[`Radio render not checked correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "pointer",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 1,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="grey"
      strokeWidth={2}
    />
  </svg>
  <input
    checked={false}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;

exports[`Radio render not checked, disabled and hover correctly 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  style={
    Object {
      "cursor": "auto",
      "height": 10,
      "outline": "none",
      "position": "relative",
      "width": 20,
    }
  }
  tabIndex={0}
>
  <svg
    baseProfile="full"
    height={20}
    style={
      Object {
        "opacity": 0.5,
        "position": "absolute",
      }
    }
    version="1.1"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx={10}
      cy={10}
      fill="#0000"
      r={8}
      stroke="grey"
      strokeWidth={2}
    />
  </svg>
  <input
    checked={false}
    disabled={true}
    id="test"
    name="test"
    onChange={[Function]}
    style={
      Object {
        "height": 20,
        "opacity": 0,
        "position": "absolute",
        "width": 20,
      }
    }
    type="radio"
    value="test"
  />
</div>
`;
